pkg/block/blocktest/adapter.go

```diff
@@ -0,0 +1,311 @@
+	reader, err := adapter.Get(ctx, dst, 0)
+	require.NoError(t, err)
+	got, err := io.ReadAll(reader)
+	require.NoError(t, err)
+	require.Equal(t, contents, string(got))
+}
+
+func testAdapterRemove(t *testing.T, adapter block.Adapter, storageNamespace string) {
+	ctx := context.Background()
+	const content = "Content used for testing"
+	tests := []struct {
+		name              string
+		additionalObjects []string
+		path              string
+		wantErr           bool
+		wantTree          []string
+	}{
+		{
+			name:     "test_single",
+			path:     "README",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+
+		{
+			name:     "test_under_folder",
+			path:     "src/tools.go",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+		{
+			name:     "test_under_multiple_folders",
+			path:     "a/b/c/d.txt",
+			wantErr:  false,
+			wantTree: []string{},
+		},
+		{
+			name:              "file_in_the_way",
+			path:              "a/b/c/d.txt",
+			additionalObjects: []string{"a/b/blocker.txt"},
+			wantErr:           false,
+			wantTree:          []string{"/a/b/blocker.txt"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// setup env
+			envObjects := tt.additionalObjects
+			envObjects = append(envObjects, tt.path)
+			for _, p := range envObjects {
+				obj := block.ObjectPointer{
+					StorageNamespace: storageNamespace,
+					Identifier:       tt.name + "/" + p,
+					IdentifierType:   block.IdentifierTypeRelative,
+				}
+				require.NoError(t, adapter.Put(ctx, obj, 0, strings.NewReader(content), block.PutOpts{}))
+			}
+
+			// test Remove
+			obj := block.ObjectPointer{
+				StorageNamespace: storageNamespace,
+				Identifier:       tt.name + "/" + tt.path,
+				IdentifierType:   block.IdentifierTypeRelative,
+			}
+			if err := adapter.Remove(ctx, obj); (err != nil) != tt.wantErr {
+				t.Errorf("Remove() error = %v, wantErr %v", err, tt.wantErr)
+			}
+
+			qk, err := adapter.ResolveNamespace(storageNamespace, tt.name, block.IdentifierTypeRelative)
+			require.NoError(t, err)
+
+			tree := dumpPathTree(t, ctx, adapter, qk)
+			if diff := deep.Equal(tt.wantTree, tree); diff != nil {
+				t.Errorf("Remove() tree diff = %s", diff)
+			}
+		})
+	}
+}
+
+func dumpPathTree(t testing.TB, ctx context.Context, adapter block.Adapter, qk block.QK) []string {
+	t.Helper()
+	tree := make([]string, 0)
+
+	uri, err := url.Parse(qk.Format())
+	require.NoError(t, err)
+
+	w, err := adapter.GetWalker(uri)
+	require.NoError(t, err)
+
+	walker := store.NewWrapper(w, uri)
+	require.NoError(t, err)
+
+	err = walker.Walk(ctx, block.WalkOptions{}, func(e block.ObjectStoreEntry) error {
+		_, p, _ := strings.Cut(e.Address, uri.String())
+		tree = append(tree, p)
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("walking on '%s': %s", uri.String(), err)
+	}
+	sort.Strings(tree)
+	return tree
+}
+
+func testAdapterMultipartUpload(t *testing.T, adapter block.Adapter, storageNamespace string) {
```

Comment:
Wonder how is this test associate to the real world case as multipart upload have minimum size of 5MB (unless it is the last part)

Reply:
AFAIK this only applies to S3.
I assume that when we enable the unit tests for S3 we will need to modify the test